Sprint 5 Feature ep-32 and ep-43 database models and pydantic update

romanvlad95 opened the pull request:
1. Jira tickets: closes [[EP-32](https://ecommercedev.atlassian.net/jira/software/projects/EP/boards/4?jql=assignee%20%3D%20712020%3Ade7c16c2-5a6b-4363-9bc4-e09d0252979e&selectedIssue=EP-32)], [[EP-43](https://ecommercedev.atlassian.net/jira/software/projects/EP/boards/4?jql=assignee%20%3D%20712020%3Ade7c16c2-5a6b-4363-9bc4-e09d0252979e&selectedIssue=EP-43)]

2. Description

This PR implements the foundational database persistence layer for the data_ingestion_service and refactors the web_scraper_service to use strongly-typed Pydantic models for content-based Kafka routing. It establishes a complete, reliable, two-service CI pipeline that validates the entire workflow, unblocking future development.

This version has been updated based on team feedback to align with the centrally-managed database schema.

3. Solution

data_ingestion_service:
- Implements SQLAlchemy ORM models for Product, Vendor, and Category, correctly mirroring the central Flyway-managed schema, including all relationships.
- Removes the local Alembic migration setup to ensure the central e-database project remains the single source of truth for schema changes.
- Adds comprehensive CRUD tests for the database models. Tests now run against a fast, in-memory SQLite database, making them self-contained and independent of any external services.
web_scraper_service:
- Adds a product_type field to all Pydantic models to enable content-based routing and data validation.
- Refactors the KafkaProducerService to publish messages to dynamic, type-specific topics (e.g., products-laptop).
- Updates the ScraperDispatcher to use a model map. Mocking it for now as it's a part of another ticket. 
CI/CD Pipeline (.github/workflows/ci.yml):
- Creates a complete, two-job CI pipeline to build and test both services in parallel.
- Simplifies and speeds up the test-data-ingestion job by removing the unnecessary PostgreSQL service container, as tests are now self-contained.
- Integrates a Kafka service container with healthchecks for reliable integration testing of the web_scraper_service.
- Configures both services and their test suites to use environment variables for all connection strings.

4. Testing
- data_ingestion_service: Unit tests for all SQLAlchemy models and CRUD operations now run against a fast, in-memory SQLite database, verifying all business logic without external dependencies.
- web_scraper_service: Unit tests for the refactored Pydantic models, the dynamic Kafka producer, and the dispatcher.
- Local Testing: All tests can be run locally for each service:


  For the web scraper
  `cd web_scraper_service && python run_tests.py`
  
  For the data ingestion service
  `cd data_ingestion_service && python run_tests.py`

- Coverage reports are generated in the htmlcov/ directory for each service.

5. Pipeline Verification

All GitHub Actions pipelines pass green. The workflow successfully spins up both Kafka and PostgreSQL services and runs the complete test suites for both microservices.
Example successful run: 
[[CI Run 1](https://github.com/MiroeCommerce/ep-WebScraper/actions/runs/17571895610)]
[[CI Run 2](https://github.com/MiroeCommerce/ep-WebScraper/actions/runs/17573593704)]
[[CI Run 3](https://github.com/MiroeCommerce/ep-WebScraper/actions/runs/17633528988)]

6. Summary

- Delivers the foundational database schema and persistence layer for the data ingestion service.
- Establishes a robust, type-safe data pipeline from the scraper to Kafka with content-based routing.
- Provides a reliable, green, multi-service CI pipeline, unblocking future team development.
- All changes ensure reproducible local runs and a stable CI process.

Next steps:
Implement the Kafka consumer in the data_ingestion_service to process messages from the new topics and save them to the database.
Begin implementing real scraping logic in the vendor-specific scraper classes to replace the current stubs.

romanvlad95 committed:
Sprint 5 feat: EP-32/EP-43 DB models, typed Pydantic models, and dynamic Kafka routing

- Implements SQLAlchemy ORM models for Product, Vendor, and Category in the data-ingestion service.
- Integrates Alembic for managing database schema migrations.
- Adds a `product_type` field to all Pydantic models in the web-scraper service to enable content-based routing.
- Refactors the KafkaProducerService to publish messages to dynamic topics based on product type (e.g., 'products-laptop').
- Updates the ScraperDispatcher to validate scraped data against the correct, type-specific Pydantic model.
- Adds a complete CI/CD pipeline in GitHub Actions for both the data-ingestion and web-scraper services.
- CI pipeline includes service dependencies (PostgreSQL, Kafka) with healthchecks for robust, reliable testing.
- Configures both services to use environment variables for all credentials and connection strings.
- Updates the entire test suite, including model, dispatcher, and end-to-end tests, to reflect the new data models and producer logic.

- All changes ensure reproducible local runs and a reliable green CI pipeline for both services.

Resolves: EP-32, EP-43

romanvlad95 committed:
fix(tests): EP-32/EP-43 Correct test suite for dispatcher and Pydantic models

- Fixes an `AttributeError` in the ScraperDispatcher's resource cleanup logic.
- Updates Pydantic model test fixtures to include the required `product_type` field, resolving multiple `ValidationErrors`.
- Corrects the end-to-end integration test to use the refactored KafkaProducerService constructor, fixing a `TypeError`.
- Refactors the API routes and scheduler to pass the necessary `category` parameter to the dispatcher.
- Removes mock code from the production dispatcher to resolve circular import issues that were breaking test collection.

- All changes ensure the entire test suite passes locally and in the CI pipeline.

Related: EP-32, EP-43

Aleks1014 commented on web_scraper_service/app/services/dispatcher.py:
```diff
@@ -1,117 +1,54 @@
-        MockScraper: An instance of the requested scraper.
-    """
-    return MockScraper()
+PRODUCT_MODEL_MAP: Dict[str, Type[BaseProduct]] = {
```
Shouldn't other product types be included as well (processor, mouse, keyboard, tablet)?

romanvlad95 replied:
I provided the dispatcher implementation as a demonstration of how the end-to-end flow works with the available models. Since you built all the product models (EP-25) you can Include them in the dispatcher file working on EP-31 and further as needed for orchestration of the scrapers logic. 

The dispatcher was provided as a helper to visualize the complete pipeline. This file is a sort of placeholder to demonstrate working e2e pipeline.

The kafka producer is designed to work with any model that follows the pattern - it discovers topics dynamically from the product_type field. While developing the consumer, any models can be tested with available ones and expand the test coverage as more models are integrated into the scraping pipeline. Thus requirements of EP-43 have been met.

metodiev commented on docker-compose.yml:
```diff
@@ -17,6 +15,16 @@ services:
+        - "5432:5432"
+      volumes:
+        - postgres_data:/var/lib/postgresql/data
 
```
Why do we need this additional PostgreDB? We already have a PostgreSQL container and project

romanvlad95 replied:
I will address this issue asap. I need to refactor the data_ingestion_service including crud file to work with the PostgreDB configuration that we have. The current implementation still uses a mocked Postgres setup. I’ll refactor it to align with the existing PostgreSQL container. As soon as I will be ready I will push necessary changes. Thx for pointing this out.

romanvlad95 replied:
Hi @metodiev.

I've pushed up the changes to address this. The data_ingestion_service has been fully refactored:
- The unnecessary PostgreSQL service has been completely removed from the docker-compose.yml.
- The service's ORM models and CRUD logic now correctly align with the central e-database schema.
- The unit tests have been updated to use a fast, in-memory SQLite database, making them self-contained and removing the need for any database service in the CI pipeline.

The latest [commit](https://github.com/MiroeCommerce/ep-WebScraper/pull/6/commits/f5c3762de9f7dd6513de55f143c99e440c5ab757) reflects this, and the CI is passing.

romanvlad95 committed:
refactor(data-ingestion): EP-32/EP-43 Align with central DB and simplify tests

- Aligns all SQLAlchemy ORM models with the centrally-managed Flyway schema.
- Removes the entire Alembic migration setup to ensure the e-database remains the single source of truth for the schema.
- Refactors the unit test suite to use a fast, in-memory SQLite database, making tests self-contained and independent of external services.
- Updates the Product model to use the generic \`sqlalchemy.JSON\` type, ensuring compatibility with both PostgreSQL and SQLite for testing.
- Simplifies the GitHub Actions CI pipeline by removing the unnecessary PostgreSQL service container, speeding up test runs.
- Resolves all database compatibility, Pydantic validation, and import errors that arose during testing.

Resolves: EP-32, EP-43
